UNION – unify types with Variant if needed

mithunputhusseri opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
UNION should unify types with Variant if needed. Resolves #82772

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

Commit:
Added fix for issue #83002

Incorporated changes to enable variant as common type for union only when use_variant_as_common_type flag is enabled

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=83246) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=83246) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=83246) it.</sub>

Avogar commented on src/Columns/getLeastSuperColumn.cpp:
```diff
@@ -36,7 +36,7 @@ ColumnWithTypeAndName getLeastSuperColumn(const std::vector<const ColumnWithType
     }
 
-    result.type = getLeastSupertype(types);
+    result.type = getLeastSupertypeOrVariant(types);
```
Let's do it under a setting too

mithunputhusseri replied:
Hi @Avogar I phrased the question wrong, This was my doubt, There was no context in this function. So I tried propogating context to this function. It required  code changes in multiple files. So I wanted to confirm with you if there was better way. Let me try things out and see what works best. Thank you

mithunputhusseri commented on src/Columns/ColumnObject.cpp:
```diff
@@ -1640,6 +1640,9 @@ void ColumnObject::takeDynamicStructureFromSourceColumns(const DB::Columns & sou
         for (const auto & [path, column_ptr] : source_object.dynamic_paths_ptrs)
         {
+            // If column_ptr is null, it means that this path is not present in the source column.
+            if (!column_ptr)
+                continue;
```
Hi @Avogar Even after adding the nullptr check, intermittently the query is failing. The query worked consistently only after adding a type_id check inside takeDynamicStructureFromSourceColumns of ColumnObject to verify SourceColumn is of ColumnObject type.

mithunputhusseri commented on src/Analyzer/UnionNode.cpp:
```diff
@@ -110,7 +110,7 @@ NamesAndTypes UnionNode::computeProjectionColumns() const
             projection_column_types[projection_index] = projections[projection_index][column_index].type;
 
-        auto result_type = getLeastSupertype(projection_column_types);
+        auto result_type = getLeastSupertypeOrVariant(projection_column_types);
```
Hi @Avogar There is context passed to this function to get the param value, can you refer to some code that I can use to get the param value

mithunputhusseri commented:
Hi @Avogar I have incorporated the review comments, Kindly review

mithunputhusseri committed:
Test fix

mithunputhusseri commented on tests/queries/0_stateless/03545_union_allow_column_with_no_common_type.sql:
```diff
@@ -0,0 +1,8 @@
+SELECT x FROM (SELECT 1 AS x, 1 AS ord UNION ALL SELECT [1,2,3] AS x, 2 AS ord) ORDER BY ord;
+SELECT x FROM (SELECT 'Hello' AS x, 1 AS ord UNION ALL SELECT map('a',1) AS x, 2 AS ord) ORDER BY ord;
+SELECT x FROM (SELECT 1 AS x, 1 AS ord UNION ALL SELECT tuple(1,'a') AS x, 2 AS ord) ORDER BY ord;
+SELECT x FROM (SELECT 1 AS x, 1 AS ord UNION ALL SELECT toUUID('00000000-0000-0000-0000-000000000000') AS x, 2 AS ord) ORDER BY ord;
```
The fix for this scenario has been added by @Avogar in this PR
https://github.com/ClickHouse/ClickHouse/pull/83295

mithunputhusseri committed:
Removed trailing white space

mithunputhusseri committed:
formatted

mithunputhusseri committed:
format change

mithunputhusseri committed:
formatted

mithunputhusseri committed:
formatted

mithunputhusseri committed:
format change

mithunputhusseri committed:
format change

mithunputhusseri committed:
fixed test

mithunputhusseri commented:
Hi @Avogar The tests I have added are failing even though I have added the parameter. I see that use_variant_as_common_type param is not getting enabled for these queries in the test even though they are specifically added. Can you let me know what is the cause?.

Avogar commented on tests/queries/0_stateless/03545_union_allow_column_with_no_common_type.sql:
You can just add `SET use_variant_as_common_type=1` in the beginning of the test instead of repeating it in every query.
Also as far as I can see, this PR implements Variant for UNION only in the new analyzer but we still have tests runs that use old analyzer and the new test is failing there: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=83246&sha=64ddd176118afb817efb66aaee41aaa0ac09c423&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+old+analyzer%2C+s3+storage%2C+DatabaseReplicated%2C+1%2F2%29&name_2=Tests

So there are 2 things we can do:
1) Implement it also with old analyzer.
2) Say that only new analyzer supports it and enable setting `enable_analyzer` in the test explicitly

Avogar commented on tests/queries/0_stateless/03545_union_allow_column_with_no_common_type.sql:
```diff
@@ -0,0 +1,8 @@
+SELECT 1 AS x UNION ALL SELECT 'Hello' ORDER BY 'x' SETTINGS use_variant_as_common_type = 1;
```
This `ORDER BY` doesn't actually order the output: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=83246&sha=64ddd176118afb817efb66aaee41aaa0ac09c423&name_0=PR&name_1=Stateless+tests+%28amd_asan%2C+flaky+check%29&name_2=Tests

mithunputhusseri committed:
merge

mithunputhusseri committed:
Incorporated review comments

mithunputhusseri commented:
Hi, Someone Kindly review the PR

Avogar committed:
Merge branch 'master' into mithunputhusseri/issue-82772

Avogar committed:
Update test 03545_union_allow_column_with_no_common_type

Avogar committed:
Update 03545_union_allow_column_with_no_common_type.reference
